### DISCUSSION TITLE
[@types/node] fs.write param options mismatch documentation

### DISCUSSION THREAD

**Discussion by wbt** [DefinitelyTyped/DefinitelyTyped#58714] [Issues with a @types package]
Type definitions for fs.write can be found [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64651581376f3b498f7d616863cfc626c8fbb11a/types/node/fs.d.ts#L2163).
[Documentation for fs.write](https://nodejs.org/api/fs.html#fswritefd-string-position-encoding-callback) in its various overloads say that the second param, titled `string`, can be either a `string` or an Object with a toString() function that returns a string. 

In my local copy, I am able to solve the issue adding the following at the top, just after the comment `Valid types for path values in "fs".` [sic]:
````typescript
    interface StringifiableObject extends Object {
        toString: () => string;
    }
    export type StringLike = string | StringifiableObject;
````
and then using `StringLike` instead of `string` as the type of the `string` parameter to the `write` functions (in multiple overloads). 
I recommend making a change like that here.

[a comment by an automated account is not shown]

**Comment by SimonSchick** [comment]
@wbt this has mostly been deliberate, implicit string conversion (just like every implicit) conversion is a common error and usually not intended.

**Comment by wbt** [comment]
@SimonSchick so there is an intentional decision that the types in DefinitelyTyped should NOT match the documentation regarding which types are accepted, and that code which is fine according to the documentation should fail due to stricter typing in this library?

What is the guiding principle about *which* type options will be prohibited by stricter typing and which will be allowed?  For example, 'path' parameters can be `PathLike = string | Buffer | URL` with implicit conversion, time parameters can be `TimeLike = string | number | Date`, etc. and those implicit conversions are seen as fine.  

As someone relatively new to this library, it would be useful to figure out when I can just rely on the official Node documentation consultations which guided creation of the code I'm converting and when I'm going to have to spend a bunch of time diving into details and finding that the type library doesn't match that documentation.  The rate at which I'm finding differences and issues with inaccurate type inference in Typescript itself really makes me think this tech isn't ready for prime time yet and the conversion is not a good idea.

**Comment by wbt** [comment]
Unless there is a larger rule about when the types should be more restrictive than Node.js official documentation, PR #58807 may fix this.
